### ISSUE TITLE
PropertiesChanged signal not received when a service starts 

### ISSUE THREAD

**Issue by cdprashanth** [systemd/systemd#570]
hi
I am writing a monitoring application to see when a service starts and exits (and take some actions on those events).
To know when a service starts I am relying on PropertiesChanged signal and query ActiveState.
But I see that PropertiesChanged signal not received when a service starts.
So its not possible to know a service start event happened.
1) Is this a bug ?
2) Is my approach of listening PropertiesChanged signal + query ActiveState correct way to monitor service start ?
Thanks

**Comment by dvdhrm** [comment]
You need to subscribe to pid1 events. Use the `Subscribe()` function on the pid1 Manager object. Once you're subscribed, you will receive notifications.

**Comment by cdprashanth** [comment]
I have already done Subscribe()  on org.freedesktop.systemd1 path =/org/freedesktop/systemd1  intf =org.freedesktop.systemd1.Manager.
But i still dont get PropertiesChanged signal not received when a service starts.

**Comment by dvdhrm** [comment]
Please provide all information for me to reproduce your bug. With your current set of information, it is impossible for me to trace your issue down.

I'd also recommend looking into other systemd utilities that already watch for these events (see logind and machined).

**Comment by cdprashanth** [comment]
Here is what I am doing:
1) Subscribe() on org.freedesktop.systemd1 path =/org/freedesktop/systemd1 intf =org.freedesktop.systemd1.Manager. to get all signals
2) systemctl stop service 
I see that I get PropertiesChanged signal
3) systemctl startservice
4) I get only JobNew, JobRemoved and UnitNew signal 
I dont get PropertiesChanged signal.
To know when a service starts I am relying on PropertiesChanged signal and query ActiveState.
Since I dont get  PropertiesChanged signal when a service starts i am finding difficult to track when a service started.

I already looked at logind and machined. did not help

```
signal sender=org.freedesktop.DBus -> dest=:1.7 serial=2 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
   string ":1.7"
signal sender=:1.1 -> dest=(null destination) serial=904 path=/org/freedesktop/systemd1; interface=org.freedesktop.systemd1.Manager; member=UnitNew
   string "cfmd.service"
   object path "/org/freedesktop/systemd1/unit/cfmd_2eservice"
signal sender=:1.1 -> dest=(null destination) serial=905 path=/org/freedesktop/systemd1; interface=org.freedesktop.systemd1.Manager; member=UnitRemoved
   string "cfmd.service"
   object path "/org/freedesktop/systemd1/unit/cfmd_2eservice"
signal sender=:1.1 -> dest=(null destination) serial=906 path=/org/freedesktop/systemd1; interface=org.freedesktop.systemd1.Manager; member=JobNew
   uint32 3417
   object path "/org/freedesktop/systemd1/job/3417"
   string "cfmd.service"
signal sender=:1.1 -> dest=(null destination) serial=907 path=/org/freedesktop/systemd1; interface=org.freedesktop.systemd1.Manager; member=JobRemoved
   uint32 3417
   object path "/org/freedesktop/systemd1/job/3417"
   string "cfmd.service"
   string "done"
signal sender=:1.1 -> dest=(null destination) serial=908 path=/org/freedesktop/systemd1; interface=org.freedesktop.systemd1.Manager; member=UnitNew
   string "cfmd.service"
   object path "/org/freedesktop/systemd1/unit/cfmd_2eservice"
```

**Comment by dvdhrm** [comment]
There is no "PropertiesChanged" signal when a unit is started. When you start a unit, the unit is already active, therefore, the 'Active' state does not change (it just gets its initial value). We do not send "PropertiesChanged" signals on object creation, only on modification.

You need to listen to "UnitNew" and then read the "Active" state to get notified about new units.

**Comment by cdprashanth** [comment]
please refer http://lists.freedesktop.org/archives/systemd-devel/2015-July/033537.html
I cant use UnitNew as in some cases we get into infinite loop
(i get unitnew on status query of non-existing unit. if i query for "Active" state i get UnitNew again and this continues.)
What is the solution in this case ?
In my programm , I want to know when a service starts

**Comment by dvdhrm** [comment]
Can you **please** provide all information you have. This is the second time you answer with data you clearly had when opening the report.

If you query systemd for dbus properties, systemd will load the unit from disk. However, as soon as no-one references this unit, anymore, it will be garbage collected. Hence, it is removed again. This also has the side effect, that reading an non-existing unit file will create the unit.

Another issues is that systemd delays notifications until the next main-loop iteration. This might coalesce multiple events into a single notifications. In the worst case, no UnitNew has been sent, yet, and the ActiveState change is coalesced into the UnitNew signal (which probably happened in your case). Maybe we should just always sent a PropertiesChanged signal alongside UnitNew. But I'll leave this up to @poettering.

If you request all properties only after you received a UnitNew event, you will not trigger a second UnitNew event. This might only happen if the unit is removed before you can send the Properties.GetAll method. Reading systemd-ui, it suffers from the same race. Maybe @keszybz has some background on this.

**Comment by cdprashanth** [comment]
Hi
Checking to see if you have any comments/updates on this
Thanks
@poettering @dvdhrm @keszybz

**Comment by poettering** [comment]
Yeah, there's indeed a nasty race: you might get a UnitNew(), then you query the property state, if the unit is already unloaded by then we will load it again, resulting in a UnitNew(), which might make the client query the property state again and so on...

I figure it's essential to ignore UnitNew() for a unit until you get the the property reply back. Or in other words: ignore the UnitNew signals for a unit if you processed a later property get msg already. Does this make sense?

**Comment by dvdhrm** [comment]
Please reopen if there're still issues. Ignoring UnitNew() while reading properties should be enough to close this race (but it might still exist if multiple processes do this query in parallel).
